0.19 Release Notes: Next Generation Scenes (#24523)
bevyengine/bevy: _release-content/release-notes/next-generation-scenes.md

---
title: "Next Generation Scenes"
authors: ["@cart"]
pull_requests: [23413, 23880, 23808, 23905, 24008]
---

**Bevy 0.19** introduces our brand new, massively improved scene system for Bevy. We've been working on this for a _long time_ (years now!), and we are excited to finally get it in the hands of Bevy developers. It makes defining scenes in code (and ultimately in assets produced by the upcoming Bevy Editor) much nicer.

### BSN (Bevy Scene Notation)

> The new scene system is modular and supports custom scene formats, but Bevy ships with its own format: BSN.

Reviewer: Weirdly soft as an opening. Maybe something like this?

```suggestion
While it's easy to hook into the internal scene machinery, we've provided a terse, clear domain-specific language for them out of the box: BSN.
```

Author: Good call. I'm thinking the next paragraph already fills this role well and supporting alternative formats is probably something we should cover later on. I'm going to cut this here, rely on the next paragraph as the intro, and add a small note that the new scene system is format-independent later on.

> BSN is an ergonomic Rust-like scene syntax which can be defined in Rust code via the `bsn!` macro _and_ in `.bsn` asset files. If you were ever bothered by the verbosity and complexity of setting up and spawning scenes in Bevy, you will probably enjoy what BSN has to offer. BSN can be used to spawn anything in the ECS, but it notably makes Bevy UI code significantly easier to read and write.

Reviewer: ```suggestion
BSN is an ergonomic Rust-like scene syntax which can be defined in Rust code via the `bsn!` macro _and_ in `.bsn` asset files. If you were ever bothered by the verbosity and complexity of setting up and spawning complex collections of entities in Bevy, you will probably enjoy what BSN has to offer. BSN can be used to spawn anything in the ECS, but it notably makes Bevy UI code significantly easier to read and write.
```

Reviewer: Most users won't recognize that they were previously spawning "scenes" in this sense.

> Note that **Bevy 0.19* supports scene assets, but we aren't yet shipping a first-party `.bsn` asset loader. **Bevy 0.19** focuses on the code-driven workflow, and we plan to roll out the asset driven workflow in the next release.

Reviewer: ```suggestion
Note that while **Bevy 0.19* supports scene assets, we aren't yet shipping a first-party `.bsn` asset loader. **Bevy 0.19** focuses on the code-driven workflow, and we plan to roll out the asset-driven workflow in the next release.
```

```rust
bsn! {
    Player
    Children [
        Sword,
        Shield,
    ]
}
```

This also works for custom relationships:
```rust
bsn! {
    Player
    Inventory [
        Apple,
        Potion,
    ]
}
```

### Scene Functions

You can define reusable BSN functions like this:
```rust
fn player() -> impl Scene {
    bsn! {
        Player
        Children [ Sword, Shield ]
    }
}
```

These can accept and use parameters:
```rust
fn player(name: &str) -> impl Scene {
    bsn! {
        Name(name)
        Player
    }
}
```

### Scenes are Composable Patches

A BSN expression is a "patch", it does not write a "full" instance of every type it defines. This means you can layer scenes on top of each other:
```rust
fn button() -> impl Scene {
    bsn! {
        Button
        Node { width: px(100) }
    }
}

fn my_button() -> impl Scene {
    bsn! {
        button()
        Node { height: px(100) }
    }
}
```

`my_button` will spawn with a `Node { width: px(100), height: px(100) }` component. Components in scenes are initialized to their defaults, and each additional scene layer writes its fields on top of those defaults.

> ### Scene Assets and Caching

Reviewer 2: the caching/assets section feels a bit out of place since end-users can't actually make use of bsn assets, etc yet. Probably better to either move it to the end of all the sections so that the other sections flow together or remove it entirely since it won't be widely usable until 0.20. (my vote is to remove it since this section also tries to commit to future work and there's plenty to talk about even without what will land for 0.20+)

Reviewer: Yeah, my preference is to cut, and move to Future Plans (people miss that!).

Author: I do see where you are coming from, but I disagree :)

"Asset scenes" and "dependency awareness" are extremely important / motivating / central parts of the scene picture. Describing theme here helps people build the picture in their mind of what this whole thing is and why we're building it. I think skipping them would be a mistake. I put the section where it is because it flows from / relates to "scenes are compose-able patches".

Also note the `:`, which is "caching" syntax. This will pre-compute the `"player.bsn"` scene and cache the results for reuse. This makes spawning much cheaper, as it only needs to resolve whatever is layered "on top" of the cached scene.

We're [working](https://github.com/bevyengine/bevy/pull/23576) on an official `.bsn` asset loader, and we also plan on porting Bevy's glTF loader to the new scene system (so you can depend on `"my_scene.gltf"` just like you would a `my_scene.bsn` file). The new scene system already supports assets, so if you're feeling adventurous you can try implementing your own Bevy scene format while you wait for ours!

### Scene Lists

`bsn!` / `Scene` corresponds to a single entity. `bsn_list!` / `SceneList` is the same idea, but applied to lists of entities:
```rust
fn players() -> impl SceneList {
    bsn_list! [
        (#Player1 Team::Blue),
        (#Player2 Team::Red),
    ]
}
```

The "BSN relationship syntax" seen above (ex: `Children []`) uses `SceneList`. This means you can pass scene lists as arguments to your scenes:
```rust
fn widget(children: impl SceneList) -> impl Scene {
    bsn! {
        Widget
        Children [ {children} ]
    }
}
```

> ### Events

Reviewer: ```suggestion
### Observers and events
```

> `bsn!` entities can easily observe events:

Reviewer: ```suggestion
`bsn!` entities can easily observe events, making it easy to embed callback-style behaviors in your scenes:
```

    Children [
        #Child Children [
            #Grandchild
        ]
    ]
}
```

In the context of `bsn_list!`, this enables defining graph structures:
```rust
bsn_list! [
    (#A PointsTo(#B)),
    (#B PointsTo(#A)),
]
```

### Implicit Into

Most values in "field position" support "implicit `.into()`". This means types that can convert into other types can generally skip manual conversion:
```rust
#[derive(Component, Default, Clone)]
struct Foo(String);

bsn! {
    Foo("hello")
}
```

This works because `"hello"` is a `&str`, which has an `Into<String>` implementation. This is especially nice in the context of defining Bevy UI values:

```rust
// Raw Rust
Node {
    border: UiRect::all(Val::Px(2.0))
    ..Default::default()
}

// BSN
Node { border: px(2) }
```

`px(2)` is just a function that produces a `Val::Px(2.0)`, and `UiRect` has an `Into` impl for `Val`, which produces `UiRect::all` (writes the value to all four border "sides"). The ergonomics here are competitive with things like CSS, but it is fully statically typed and derived from normal Rust trait conversions (these aren't special cased / hard-coded). This means you can build your own!

### Scene Components

> BSN makes it possible to associate a `Scene` with a `Component` via the `SceneComponent` derive:

Reviewer: These needs more motivation.

Scene Components can also define "props" which are passed into the scene function and can inform BSN outputs:
```rust
#[derive(SceneComponent, Default, Clone)]
#[scene(PlayerProps)]
struct Player {
    score: usize,
}

#[derive(Default)]
struct PlayerProps {
    alignment: Alignment
}

impl Player {
    fn scene(props: PlayerProps) -> impl Scene {
        let alignment: Box<dyn Scene> = match props.alignment {
            Alignment::Good => Box::new(bsn! { AngelWings }),
            Alignment::Evil => Box::new(bsn! { DevilHorns }),
        };
        bsn! {
            #Player
            alignment
            Items [ Sword, Shield ]
        }
    }
}

bsn! {
    @Player {
        @alignment: Alignment::Good,
        score: 10,
    }
}
```

> ### Scene Spawn Systems

Reviewer: ```suggestion
### Scene Spawning Systems
```

> **Bevy 0.19** ships with helpers to easily spawn scene functions. This is a _fully self-contained_ Bevy app that spawns a 3D scene:

Reviewer: "helpers" is not very useful when you only demonstrate one.

```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, level.spawn())
        .run();
}

> fn level() -> impl SceneList {

Reviewer: A comment to the effect of ".spawn turns any function which returns a scene into a system" would be good.

Author: Good call. I'm making it a followup sentence because I want to keep the code block as small and free of context as possible to help drive the point home. I'm also porting the 3D example to a 2D example to further simplify things to achieve that goal.

```rust
bsn! {
    Player
    Children [
        Sword,
        Shield,
    ]
}
```

This also works for custom relationships:

```rust
bsn! {
    Player
    Inventory [
        Apple,
        Potion,
    ]
}
```

### Scene Functions

You can define reusable BSN functions like this:

```rust
fn player() -> impl Scene {
    bsn! {
        Player
        Children [ Sword, Shield ]
    }
}
```

These can accept and use parameters:

```rust
fn player(name: &str) -> impl Scene {
    bsn! {
        Name(name)
        Player
    }
}
```

### Scenes are Composable Patches

A BSN expression is a "patch", it does not write a "full" instance of every type it defines. This means you can layer scenes on top of each other:

```rust
fn button() -> impl Scene {
    bsn! {
        Button
        Node { width: px(100) }
    }
}

fn my_button() -> impl Scene {
    bsn! {
        button()
        Node { height: px(100) }
    }
}
```

`my_button` will spawn with a `Node { width: px(100), height: px(100) }` component. Components in scenes are initialized to their defaults, and each additional scene layer writes its fields on top of those defaults.

### Scene Assets and Caching

> While **Bevy 0.19** doesn't ship with an official `.bsn` asset loader, it _does_ already support scene asset dependencies:

Reviewer 3: This isn't very clear, sounds like we're saying "it does and doesn't support scene assets" which seems confusing. ofc *I* know what you're trying to say, but i would imagine others wouldn't.

Author: I'll add some clarity here.

```rust
commands.queue_spawn_scene(bsn! {
    :"player.bsn"
    Transform {
        translation: Vec3 { x: 10. }
    }
})
```

> This will spawn a scene that includes the `"player.bsn"` scene asset and patches the "x position" to be `10`. BSN is dependency-aware: if you use `queue_spawn_scene` instead of `spawn_scene`, it will wait to spawn the scene until all dependencies have loaded. `spawn_scene` will always try to spawn the scene immediately ... if it has scene dependencies that aren't loaded yet it will fail.

Reviewer 4: If there's no asset loader, how is the scene asset included? Is this literally calling `include!()` behind the scenes? How is path resolution? Is the caching at run-time, compile time?

Author: I'll add a bit more clarity here

Also note the `:`, which is "caching" syntax. This will pre-compute the `"player.bsn"` scene and cache the results for reuse. This makes spawning much cheaper, as it only needs to resolve whatever is layered "on top" of the cached scene.

> We're [working](https://github.com/bevyengine/bevy/pull/23576) on an official `.bsn` asset loader, and we also plan on porting Bevy's glTF loader to the new scene system (so you can depend on `"my_scene.gltf"` just like you would a `my_scene.bsn` file). The new scene system already supports assets, so if you're feeling adventurous you can try implementing your own Bevy scene format while you wait for ours!

Reviewer 3: Again with the does or doesn't support. Perhaps it makes sense to say "the syntax and spawning code already supports..." instead of "the scene system already supports..." to more directly help people understand whats missing

Author: Yeah being slightly more specific here makes sense.

### Scene Lists

`bsn!` / `Scene` corresponds to a single entity. `bsn_list!` / `SceneList` is the same idea, but applied to lists of entities:

```rust
fn players() -> impl SceneList {
    bsn_list! [

>         (#Player1 Team::Blue),

Reviewer 4: Is the parenthesis here special in any way? Is it a `Bundle`? is it for disambiguation/clarity or required syntax?

Would something like:

 ```
bsn_list![
    #Player1 Team::Blue,
    #Player2 Team::Red,
]
```

work just as well?

The "old" approach to spawning via bundles required passing in every ECS dependency into a bundle function and manually using that dependency to produce the final value:

```rust
fn player(asset_server: &AssetServer) -> impl Bundle {
    (
        Player {
            score: 10,
            ..Default::default()
        },
        children! [
            Sprite {
                image: asset_server.load("player.png"),
                ..Default::default()
            }
        ]
    )
}

fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
    commands.spawn(player(&asset_server))
}
```

This gets _quite_ nasty when spawning complex deeply nested scenes with many dependencies.

BSN makes this all much easier:

```rust
fn player() -> impl Scene {
    bsn! {
        Player { score: 10 }
        Children [
            Sprite { image: "player.png" }
        ]
    }
}

fn setup(mut commands: Commands) {
    commands.spawn_scene(player());
}
```

Spawning a scene no longer requires knowing every little dependency it requires internally, and common actions like loading and assigning assets via their paths is simple!

> This does mean that BSN requires types to have a `Template`. This is accomplished via the `FromTemplate` trait, which tells BSN what `Template` type it should use for a given `Component`. `FromTemplate` can be derived, which will also generate a `Template` type for your type. Fortunately, most types _do not_ need to derive or implement `FromTemplate` manually. This is because `FromTemplate` and `Template` is automatically implemented for every type that implements `Default` and `Clone`. These types are "templates of themselves" and are just "passed through". You only need to derive `FromTemplate` if you need template features (such as the `Sprite` use case above, which uses a `Handle<Image>` template to accept `"player.png"`).

Reviewer 3: "to have a `Template`" sounds odd, i'd go with "to implement Template"

Reviewer 5: Personally disagree, I think the concept of components "having a template" is not only intuitive, it's important for the mental model

Author: Yup "having a Template" is the correct mental model here. This is the FromTemplate vs Template distinction.

> ### Inline Asset Templates

Reviewer 3: Since this asset_value is mentioned, should template_value also be mentioned somewhere in here? its release notes, not full docs...

Author: I thought about this: I think skipping this saves us space without costing us too much. I agree that these aren't docs, just a feature overview. We don't need to fully specify everything.

    Children [
        #Child Children [
            #Grandchild
        ]
    ]
}
```

In the context of `bsn_list!`, this enables defining graph structures:

```rust
bsn_list! [
    (#A PointsTo(#B)),
    (#B PointsTo(#A)),
]
```

### Implicit Into

Most values in "field position" support "implicit `.into()`". This means types that can convert into other types can generally skip manual conversion:

```rust
#[derive(Component, Default, Clone)]
struct Foo(String);

bsn! {
    Foo("hello")
}
```

This works because `"hello"` is a `&str`, which has an `Into<String>` implementation. This is especially nice in the context of defining Bevy UI values:

```rust
// Raw Rust
Node {
    border: UiRect::all(Val::Px(2.0))
    ..Default::default()
}

// BSN
Node { border: px(2) }
```

`px(2)` is just a function that produces a `Val::Px(2.0)`, and `UiRect` has an `Into` impl for `Val`, which produces `UiRect::all` (writes the value to all four border "sides"). The ergonomics here are competitive with things like CSS, but it is fully statically typed and derived from normal Rust trait conversions (these aren't special cased / hard-coded). This means you can build your own!

### Scene Components

> It has almost been a Bevy developer right of passage to define something like a `Player` component, which has complex behaviors that rely on some larger "scene", and then ask questions like "how to I spawn this all together?" and "how do I write code that can safely assume the whole scene is present?". Bevy developers have solved these problems in a variety of creative ways, but there has never been an easy recommended / idiomatic upstream solution.

Reviewer 3: A bit misleading: after spawning, you *can't* rely on the assumption the components *stay* this way. You can only assume they are *spawned* this way. Very different guarantee.

BSN solves this problem by making it possible to associate a `Scene` with a `Component` via the `SceneComponent` derive:

```rust
#[derive(SceneComponent, Default, Clone)]
struct Player {
    score: usize
}

impl Player {
    fn scene() -> impl Scene {
        bsn! {
            Transform { translation: Vec3 { x: 10. } }
            Children [
                LeftHand,
                RightHand,
            ]
        }
    }
}
```

Scene components can then be spawned like this:

```rust
world.spawn_scene(bsn! {
    @Player { score: 10 }
})
```

> Scene Components must be spawned this way (as a "scene component"), and will log errors if they are spawned directly (ex: via `world.spawn(Player::default())`). Critically, this provides the guarantee that if the `Player` component is present, the full scene will also be present. As a developer this means you can write code that queries for `Player` and assume that it will have both a `LeftHand` and a `RightHand` child. This was a major missing piece in the Bevy data model!

Reviewer 4: ```suggestion
Scene Components must be spawned this way (as a "scene component"), and will log errors if they are spawned directly (ex: via `world.spawn(Player::default())`). Critically, this provides the guarantee that if the `Player` component is present, the full scene will also be present at spawn time. As a developer this means you can write code that queries for `Player` and assume that it will have both a `LeftHand` and a `RightHand` child. This was a major missing piece in the Bevy data model!
```

If I understand it correctly, you can still remove/despawn things, it doesn't make these always guaranteed to exist, right?

Scene Components can also define "props" which are passed into the scene function and can inform BSN outputs:

```rust
#[derive(SceneComponent, Default, Clone)]
#[scene(PlayerProps)]
struct Player {
    score: usize,
}

#[derive(Default)]
struct PlayerProps {
    alignment: Alignment
}

impl Player {
    fn scene(props: PlayerProps) -> impl Scene {
        let alignment: Box<dyn Scene> = match props.alignment {
            Alignment::Good => Box::new(bsn! { AngelWings }),
            Alignment::Evil => Box::new(bsn! { DevilHorns }),
        };
        bsn! {
            #Player
            alignment
            Items [ Sword, Shield ]
        }
    }
}

bsn! {
    @Player {
        @alignment: Alignment::Good,

>         score: 10,

Reviewer 4: It's in the example, but might make sense to explicitly call out that props and fields can be mixed?

Also what the "order" in which the field / prop "wins" when evaluating the BSN? (Is it the same as the one mentioned for composable patch behavior? Or are fields different?)

    }
}
```

> The `SceneComponent` derive also supports shorthand for scene assets:

Reviewer 3: not just that, the same syntax also accepts functions

    Children [
        #Child Children [
            #Grandchild
        ]
    ]
}
```

In the context of `bsn_list!`, this enables defining graph structures:

```rust
bsn_list! [
    (#A PointsTo(#B)),
    (#B PointsTo(#A)),
]
```

## Implicit Into

Most values in "field position" support "implicit `.into()`". This means types that can convert into other types can generally skip manual conversion:

```rust
#[derive(Component, Default, Clone)]
struct Foo(String);

bsn! {
    Foo("hello")
}
```

This works because `"hello"` is a `&str`, which has an `Into<String>` implementation. This is especially nice in the context of defining Bevy UI values:

```rust
// Raw Rust
Node {
    border: UiRect::all(Val::Px(2.0))
    ..Default::default()
}

// BSN
Node { border: px(2) }
```

`px(2)` is just a function that produces a `Val::Px(2.0)`, and `UiRect` has an `Into` impl for `Val`, which produces `UiRect::all` (writes the value to all four border "sides"). The ergonomics here are competitive with things like CSS, but it is fully statically typed and derived from normal Rust trait conversions (these aren't special cased / hard-coded). This means you can build your own!

## Scene Components

> It has almost been a Bevy developer rite of passage to define something like a `Player` component, which has complex behaviors that rely on some larger "scene", and then ask questions like "how do I spawn this all together?" and "how do I write code that can safely assume the whole scene is present?". Bevy developers have solved these problems in a variety of creative ways, but there has never been an easy recommended / idiomatic upstream solution.

Reviewer 6: Maybe it's a me problem, but I got confused by this section because I didn't understand that it was about spawning *separate entities* together, not just components (which would be a lot easier). It's obvious looking back, but now I'm thinking, if I got confused as someone relatively familiar with Bevy, I can definitely see less-familiar people getting confused.

Reviewer 4: I think the issue might be because of the specific example, since hands are part of the player, not children.

Maybe something that more clearly is a child entity (e.g. Car with Driver and Passenger) or something like that

Author: > since hands are part of the player, not children

I would argue that hands should be their own related entities (ex: children). They might each have their own slots, some players might have 1 hand, others might have 4, etc.

I see where this is coming from, but I'm not sure its worth changing. I like using the `Player` entity in examples as its somewhat "universal" / easy to relate to.

Reviewer 4: Fair, but I meant it more mesh-wise

Maybe Sword and Shield? 🗡️🛡️ ;)

> It has almost been a Bevy developer right of passage to define something like a `Player` component, which has complex behaviors that rely on some larger "scene", and then ask questions like "how to I spawn this all together?" and "how do I write code that can safely assume the whole scene is present?". Bevy developers have solved these problems in a variety of creative ways, but there has never been an easy recommended / idiomatic upstream solution.

Reviewer 7: Very exciting so far! Small typo here: shouldn't it be `rite of passage` instead of `right of passage`?

Reviewer 3: yeah, its "rite of passage"

Reviewer 7: "how to I spawn this all together?" -> "how **d**o I spawn this all together?"
